From 8d0542425385cce2b1e63f38b0eea27f6657a0f7 Mon Sep 17 00:00:00 2001
From: anonymous <anonymous@inventec.com>
Date: Tue, 29 Sep 2020 19:26:24 -0400
Subject: [PATCH] Subject: [Patch][kernel][RTC] Set a default timestamp for rtc
 init failed condition

- If the RTC value read from register is invalid then set it to default timestamp
---
 drivers/rtc/class.c | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/drivers/rtc/class.c b/drivers/rtc/class.c
index f77bc089eb6b..ecebb03f676f 100644
--- a/drivers/rtc/class.c
+++ b/drivers/rtc/class.c
@@ -36,6 +36,8 @@ static void rtc_device_release(struct device *dev)
 /* Result of the last RTC to system clock attempt. */
 int rtc_hctosys_ret = -ENODEV;
 
+#define DEFAULT_RTC_TIMESTAMP (1598918400) // Define a default timestamp 2020-09-01 Tuesday 00:00:00
+
 /* IMPORTANT: the RTC only stores whole seconds. It is arbitrary
  * whether it stores the most close value or the value with partial
  * seconds truncated. However, it is important that we use it to store
@@ -59,7 +61,14 @@ static void rtc_hctosys(struct rtc_device *rtc)
 	if (err) {
 		dev_err(rtc->dev.parent,
 			"hctosys: unable to read the hardware clock\n");
-		goto err_read;
+		// goto err_read;
+
+		// Due to some platform doesn't have battery to keep RTC register count
+		// once AC lost, the value of RTC register is unknown
+		// that will cause system time init failed.
+		// So we set a default timestamp to avoid this case.
+		rtc_time64_to_tm(DEFAULT_RTC_TIMESTAMP, &tm);
+		rtc_set_time(rtc, &tm);
 	}
 
 	tv64.tv_sec = rtc_tm_to_time64(&tm);
-- 
2.17.1

